### PR TITLE
remove `"-Xmigration"` from main scalacOptions. add explicit type

### PR DESCRIPTION
**Pull request by xuwei-k**
# Pull Request Checklist

* [x] Have you read [How to write the perfect pull request](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/)?
* [x] Have you read through the [contributor guidelines](https://www.playframework.com/contributing)?
* [x] Have you referenced any issues you're fixing using [commit message keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
* [x] Have you added copyright headers to new files?
* [x] Have you checked that both Scala and Java APIs are updated?
* [x] Have you updated the documentation for both Scala and Java sections?
* [x] Have you added tests for any changed functionality?

# Helpful things

## Purpose

remove `"-Xmigration"` from main code and add explicit types. for avoid different return types Scala 2 and 3.

## Background Context

`"-Xmigration"` suppress some warnings. But I think this is **not** good idea.

please see following example and `scala/scala` pull request.

- https://github.com/scala/scala/pull/10439

#### Scala 3

```scala
Welcome to Scala 3.3.1 (11.0.20.1, Java OpenJDK 64-Bit Server VM).
Type in expressions for evaluation. Or try :help.
                                                                                                                                             
scala> trait A { def x: Any } ; class B extends A { def x = "x" }
// defined trait A
// defined class B
                                                                                                                                             
scala> new B().x // not String !. different type with Scala 2.13
val res0: Any = x

```

#### no option

```scala
Welcome to Scala 2.13.12 (OpenJDK 64-Bit Server VM, Java 11.0.20.1).
Type in expressions for evaluation. Or try :help.

scala> trait A { def x: Any } ; class B extends A { def x = "x" }
trait A
class B

scala> new B().x
val res0: String = x
```

#### `-Xsource:3`

```scala
$ scala -Xsource:3
Welcome to Scala 2.13.12 (OpenJDK 64-Bit Server VM, Java 11.0.20.1).
Type in expressions for evaluation. Or try :help.

scala> trait A { def x: Any } ; class B extends A { def x = "x" }
                                                        ^
       error: under -Xsource:3, inferred Any instead of String [quickfixable]
       Scala 3 migration messages are errors under -Xsource:3. Use -Wconf / @nowarn to filter them or add -Xmigration to demote them to warnings.
       Applicable -Wconf / @nowarn filters for this fatal warning: msg=<part of the message>, cat=scala3-migration, site=B.x

scala> trait A { def x: Any } ; class B extends A { def x: String = "x" }
trait A
class B
```

#### `-Xsource:3` and `-Xmigration`

```scala
$ scala -Xsource:3 -Xmigration
Welcome to Scala 2.13.12 (OpenJDK 64-Bit Server VM, Java 11.0.20.1).
Type in expressions for evaluation. Or try :help.

scala> trait A { def x: Any } ; class B extends A { def x = "x" }
                                                        ^
       warning: under -Xsource:3, inferred Any instead of String [quickfixable]
trait A
class B

scala> def f: Seq[Int] = Nil // unnecessary migration message!!!
              ^
       warning: type Seq in package scala has changed semantics in version 2.13.0:
       scala.Seq is now scala.collection.immutable.Seq instead of scala.collection.Seq
def f: Seq[Int]
```

### CODE REVIEW

**Review by ihostage — APPROVED** [review]

**Comment by mkurz** [comment]
@xuwei-k can you please rebase? Thanks!

**Review by mkurz — APPROVED** [review]
Thanks LGTM!

However, I pushed two commits:
- First I removed the MiMa filter you added because it is not necessary (anymore?)
- I also removed the `-Wconf...` flags you added, because also they are not necessary anymore because Scala 2.13.13 fixed the errors, see 
  - https://github.com/scala/scala-dev/issues/441#issuecomment-1750654658
  - https://github.com/scala/bug/issues/12886
  - https://github.com/scala/scala/pull/10573 <-- This PR removes the `package object inheritance is deprecated` errors/warnings

File: project/PlayBuildBase.scala [hunk 1]
```diff
@@ -56,9 +56,17 @@ object PlayBuildBase extends AutoPlugin {
...
         case _             => Seq.empty
       }),
+    Test / scalacOptions ++= {
+      CrossVersion.partialVersion(scalaVersion.value) match {
+        case Some((2, 13)) =>
+          Seq("-Xmigration:2.13")
+        case _ =>
+          Seq.empty
+      }
+    },
```
Comment (mkurz) [inline project/PlayBuildBase.scala:69]: The `-Xmigration` flag for the test scope eventually got removed in
- #12577

I adjusted all the test finally (necessary for upgrading to Scala 2.13.14 without adding any `Wconf` filters, see https://github.com/playframework/playframework/pull/12576#issuecomment-2085336292)

**Comment by mkurz** [comment]
Backported this changes (together with removing `-Xmigration` from tests) to 2.9.x and 3.0.x:
- #12576
- #12575

**Pull request merged**